### ISSUE TITLE
#as_json option :only Not working

### ISSUE THREAD

**Issue by brosintoski** [rails-api/active_model_serializers#1851]
#### Expected behavior vs actual behavior
Expected: Only fields passed to only option are returned from `as_json`.

Actual:  All fields/attributes of the serializer are returned.

#### Steps to reproduce
1.  Create an activerecord model called User with say name and status.
2.  Create a serializer for that model with `attributes :id, :name, :status`
3.  From console execute following steps:
     `UserSerializer.new(User.new(name: 'foo bar', status: 'active')).as_json(only: [:id, :name]`)
4.  Note that status is also returned.

#### Environment

ActiveModelSerializers Version *(commit ref if not on tag)*: aa4d89ab4765aa7721aa053840a4bb57152ea7ba

Output of `ruby -e "puts RUBY_DESCRIPTION"`: ruby 2.3.1p112 (2016-04-26 revision 54768) [x86_64-darwin15]

OS Type & Version: OS X v10.11.5

Integrated application and version *(e.g., Rails, Grape, etc)*: rails 4.2.6

#### Backtrace
*(e.g., provide any applicable backtraces from your application)*

#### Additonal helpful information
*(e.g., Gemfile.lock, configurations, PR containing a failing test, git bisect results)*

**Comment by bf4** [comment]
So, this isn't currently supported.  Are there docs that suggest otherwise?  I've been thinking about the differences between the AMS serializer api and the active model serialization api, but nothing's happened yet.

Related #1845

**Comment by brosintoski** [comment]
@bf4 yes yard doc examples found here: 

- https://github.com/rails-api/active_model_serializers/blob/master/lib/active_model/serializer.rb#L155
- http://www.rubydoc.info/github/rails-api/active_model_serializers/ActiveModel/Serializer#serializable_hash-instance_method

**Comment by bf4** [comment]
@brosintoski Well, that's clearly a major fail on my part. Those comments weren't meant to stick around so long.  

```plain
    # TODO: Include <tt>ActiveModel::Serializers::JSON</tt>.
    # So that the below is true:
````

was meant for people reading the code to see that this should be implemented...  I think if I just made it an issue it would get lost.. hmm

**Comment by brosintoski** [comment]
@bf4 no problem.  For now I just override `#relationship_value_for` in my base serializer to allow support for something like:  `fields: {roles: [:id, :name], default_account: [:id, :code]}`.  To me that format is more inline with arel include option.  The nested only/exclude just add unnecessary complexity since I never use exclude logic.

With that said I'm open to any design that achieves filtering of base and associative attributes as a serialization option, so will just watch this project for updates.

**Comment by bf4** [comment]
Good eyes on `relationship_value_for`.  That method is does a bit too much.  I'm working in it right now to fix https://github.com/rails-api/active_model_serializers/pull/1857

**Comment by bf4** [comment]
@brosintoski Mind include a snippet for anyone else whom it might help?  (Or maybe it should even be a PR?)

**Comment by brosintoski** [comment]
Happy to share the code but would consider it a stop gap solution for now due to how `relationship_value_for` is structured.  Had to completely override that method (i.e. no use of super).

Here is the code:

```
  def options_for_association(association,adapter_options)
    options = {}
    fields_option = if !adapter_options.key?(:fields)
      nil
    elsif adapter_options[:fields].is_a?(Hash)
      adapter_options[:fields][association.name]
    elsif adapter_options[:fields].is_a?(Array)
      detected_fields_option = adapter_options[:fields].detect {|x| x[association.name] if x.is_a?(Hash) && x.key?(association.name) }
      detected_fields_option[association.name] if detected_fields_option
    end
    options.merge!(fields: fields_option) if fields_option
    options
  end

  def relationship_value_for(association, adapter_options, adapter_instance)
    return association.options[:virtual_value] if association.options[:virtual_value]
    association_serializer = association.serializer
    association_object = association_serializer && association_serializer.object
    return unless association_object

    options = options_for_association(association, adapter_options) || {}
    relationship_value = association_serializer.serializable_hash(adapter_options, options, adapter_instance)

    if association.options[:polymorphic] && relationship_value
      polymorphic_type = association_object.class.name.underscore
      relationship_value = { type: polymorphic_type, polymorphic_type.to_sym => relationship_value }
    end

    relationship_value
  end
```

**Comment by bf4** [comment]
@brosintoski yeah, method just part of an ongoing effort to improve the design, which is why I labeled it as api private :)

**Comment by bf4** [comment]
Related https://github.com/rails-api/active_model_serializers/issues/1643
